compiler/next/lib/resolution/resolution-queries.cpp

```diff
@@ -33,313 +35,1441 @@
+    return nullptr;
+  }
+
+  // TODO: check method-ness
+  // TODO: reason failed
+
+  auto initialTypedSignature = typedSignatureInitial(context, uSig);
+  // Next, check that the types are compatible
+  size_t formalIdx = 0;
+  for (const FormalActual& entry : faMap.byFormalIdx) {
+    const auto& actualType = entry.actualType;
+    const auto& formalType = initialTypedSignature->formalTypes[formalIdx];
+    bool ok = canPassInitial(actualType, formalType);
+    if (!ok) {
+      return nullptr;
+    }
+
+    formalIdx++;
+  }
+
+  // check that the where clause applies
+  if (initialTypedSignature->whereClauseResult==TypedFnSignature::WHERE_FALSE) {
+    return nullptr;
   }
 
-  // Delve further into the children of the node
-  for (const ASTNode* child : ast->children()) {
-    resolveAST(context, child, useScope, resultByPostorderID, undefined);
+  return initialTypedSignature;
+}
+
+// returns nullptr if the candidate is not applicable,
+// or the result of an instantiated typedSignature if it is.
+static const TypedFnSignature*
+doIsCandidateApplicableInstantiating(Context* context,
+                                     const TypedFnSignature* typedSignature,
+                                     const CallInfo& call,
+                                     const PoiScope* poiScope) {
+
+  const TypedFnSignature* instantiated =
+    instantiateSignature(context, typedSignature, call, poiScope);
+
+  if (instantiated == nullptr)
+    return nullptr;
+
+  // Next, check that the types are compatible
+  size_t nActuals = call.actuals.size();
+  for (size_t i = 0; i < nActuals; i++) {
+    const QualifiedType& actualType = call.actuals[i].type;
+    const QualifiedType& formalType = instantiated->formalTypes[i];
+    bool ok = canPass(actualType, formalType);
+    if (!ok)
+      return nullptr;
   }
+
+  // check that the where clause applies
+  if (instantiated->whereClauseResult == TypedFnSignature::WHERE_FALSE)
+    return nullptr;
+
+  return instantiated;
 }
 
-const ResolvedSymbol&
-resolveModule(Context* context, const Module* mod) {
-  QUERY_BEGIN(resolveModule, context, mod);
+static const TypedFnSignature* const&
+isCandidateApplicableInitialQuery(Context* context,
+                                  ID candidateId,
+                                  CallInfo call) {
 
-  ResolutionResultByPostorderID resolutionById;
-  std::set<UniqueString> undefined;
-  resolveAST(context, mod, nullptr, resolutionById, undefined);
+  QUERY_BEGIN(isCandidateApplicableInitialQuery, context, candidateId, call);
 
-  ResolvedSymbol result;
-  result.decl = mod;
-  result.resolutionById.swap(resolutionById);
+  const TypedFnSignature* result =
+    doIsCandidateApplicableInitial(context, candidateId, call);
 
   return QUERY_END(result);
 }
 
-const ResolvedSymbolVec& resolveFile(Context* context, UniqueString path) {
-  QUERY_BEGIN(resolveFile, context, path);
+const std::vector<const TypedFnSignature*>&
+filterCandidatesInitial(Context* context,
```

Comment:
When writing the scope resolution code, I created `BorrowedIdsWithName` that tries to be efficient by pointing to a list of the IDs with a given name in a given scope. That led to this situation where there might be a `moreIds` or there might not be, because that's how that data type works. It's arguably an optimization though (and we could consider using a 1-long 'moreIds' if there is exactly 1 candidate).

Anyway, yes, `lst` has 1 entry per scope where there was a match. Within that entry we might have just 1 match or many (if there are several overloads, say).